runtime/memprof.c

```diff
@@ -156,1176 +1258,983 @@ static memprof_domain_t domain_create(caml_domain_state *caml_state)
-    return Invalid_index;
-  local->entries.len++;
-  t = &local->entries.t[local->entries.len - 1];
-  t->block = block;
-  t->n_samples = n_samples;
-  t->wosize = wosize;
-  t->user_data = user_data;
-  t->running = NULL;
-  t->alloc_young = is_young;
-  t->source = source;
-  t->promoted = 0;
-  t->deallocated = 0;
-  t->cb_promote_called = t->cb_dealloc_called = 0;
-  t->deleted = 0;
-  return local->entries.len - 1;
+  value res = callstack;
+  if (Is_long(callstack)) {
+    callstack_stash_t stash = Ptr_val(callstack);
+    res = caml_alloc(stash->frames, 0);
+    for (size_t i = 0; i < stash->frames; ++i) {
+      Store_field(res, i, Val_backtrace_slot(stash->stack[i]));
+    }
+    caml_stat_free(stash);
+  }
+  return res;
 }
 
-static void mark_deleted(struct entry_array* ea, uintnat t_idx)
+/**** Running callbacks ****/
+
+/* Runs a single callback, in thread `thread`, for entry number `i` in
+ * table `es`. The callback closure is `cb`, the parameter is `param`,
+ * and the "callback index" is `cb_index`.
+ * Returns Val_unit or an exception result.
+ * Updates `*moved_p` to `true` if the entry is moved to a new table
+ * during the callback (for instance, if the callback - or another
+ * thread of this domain - stops the profile and starts a new one). */
+
+static value run_callback_exn(memprof_thread_t thread,
+                              entries_t es, size_t i,
+                              value cb, value param,
+                              uintnat cb_index,
+                              bool *moved_p)
 {
-  struct tracked* t = &ea->t[t_idx];
-  t->deleted = 1;
-  t->user_data = Val_unit;
-  t->block = Val_unit;
-  if (t_idx < ea->delete_idx) ea->delete_idx = t_idx;
-}
+  entry_t e = &es->t[i];
+  value res = Val_unit;
+  bool moved = false;
+
+  if (e->running) { /* some other thread has got to this callback first */
+    goto done;
+  }
+
+  thread->running_table = es;
+  thread->running_index = i;
+  e->running = thread;
+
+  e->callback = cb_index;
+  e->callbacks |= CB_MASK(cb_index);
+  e->user_data = Val_unit;      /* Release root. */
 
-Caml_inline value run_callback_exn(
-  struct entry_array* ea, uintnat t_idx, value cb, value param)
-{
-  struct tracked* t = &ea->t[t_idx];
-  value res;
-  CAMLassert(t->running == NULL);
-  CAMLassert(lambda > 0.);
-
-  local->callback_status = ea == &entries_global ? t_idx : CB_LOCAL;
-  t->running = local;
-  t->user_data = Val_unit;      /* Release root. */
   res = caml_callback_exn(cb, param);
-  if (local->callback_status == CB_STOPPED) {
-    /* Make sure this entry has not been removed by [caml_memprof_stop] */
-    local->callback_status = CB_IDLE;
-    return Is_exception_result(res) ? res : Val_unit;
+  /* The entry may have been moved to another table under our feet,
+   * due to the callback or to other threads from this domain. For
+   * example, if a new profile is started. */
+  entries_t new_es = thread->running_table;
+  moved = (es != new_es);
```

Comment:
```suggestion
  *moved_p = (es != new_es);
```

No need for a local variable `moved` here.